### DISCUSSION TITLE
Semantic-release permanently broken (reading the incorrect version), cannot get it to release a new version

### DISCUSSION THREAD

**Discussion by paambaati** [semantic-release/semantic-release#2788] [Q&A]
Many months ago, I migrated the default branch of my repo from `master` to `main`, and ever since then, semantic-release has never been the same.

I keep going back to the troubleshooting guide and do all the steps, but the release process is now permanently broken and has become unreliable when applying manual fixes.

Example workflow run that reports the wrong version – https://github.com/paambaati/codeclimate-action/actions/runs/4905711649/jobs/8759663416

The dry run keeps picking up the last commit as `v3.2.0` when in fact, `v4.0.0` was released by semantic-release (see https://github.com/paambaati/codeclimate-action/actions/runs/4777730343/jobs/8493671186). I've also followed the troubleshooting guide to recreate my tags and ref notes, with no luck.

```bash
# REFER: https://github.com/semantic-release/semantic-release/blob/master/docs/support/troubleshooting.md#release-not-found-release-branch-after-git-push---force
git tag --delete v4.0.0
git push --delete origin v4.0.0
git tag v4.0.0 cdbb31e9cece505753becbfccc7ab64665a4cca5 # Commit that created the release
git notes --ref semantic-release add -f -m '{"channels":[null]}' v4.0.0
git push --force origin refs/notes/semantic-release
```

EDIT: Additional information as I continue to debug this –

Looks like `git tag --merged main` returns only a few tags.

```bash
git tag --merged main

v1.0.0
v2.2.5
v2.2.6
v3.1.0
```

The above steps no longer work either, and I'm at my wits' end. I'd really appreciate if someone can help fix this for good.

**Comment by travi** [comment]
> Looks like `git tag --merged main` returns only a few tags.

if the previous tag does not exist in the branch you are releasing from, it will not be found by semantic-release. the tag needs to exist _and_ it need to be in the history of the branch you are releasing from. if it is there properly, semantic-release does work. also, if you are creating tags locally in an attempt to resolve an issue with missing tags, be sure you are actually pushing the tags to your remote. that normally requires a `--tags` or `--follow-tags` flag when pushing.

**Comment by paambaati** [comment]
**EDIT:** The below solution worked only once.

I was able to solve this by doing these steps –

1. Say v4.0.0 is the latest release tag (but the one that isn't being detected by semantic-release); the fix initially is to force-redo the tag –
    ```bash
    # delete the tags first (on remote too)
    git push origin :refs/tags/v4.0.0
    git push origin :refs/tags/v4.0
    git push origin :refs/tags/v4
    # recreate the tag at the correct commit (i.e. the commit that triggered the original v4.0.0 release tag)
    # (you can get the commit for the tag from the releases page)
    git tag -fa v4.0.0 bcd27f6 -m "Release v4.0.0"
    git tag -fa v4.0 bcd27f6 -m "Release v4.0.0"
    git tag -fa v4 bcd27f6 -m "Release v4.0.0"
    # add ref notes to make it look like semantic-release created these tags
    git notes --ref semantic-release add -f -m '{"channels":[null]}' v4.0.0
    git notes --ref semantic-release add -f -m '{"channels":[null]}' v4.0
    git notes --ref semantic-release add -f -m '{"channels":[null]}' v4
    ```
2. Push these to the remote –
    ```bash
    git push origin --tags
    git push --force origin refs/notes/semantic-release
    ```
3. Trigger the workflow to have semantic-release pick up the new version that comes after v4.0.0.

**Reply by gr2m** [reply]
thanks for sharing your solution!

**Reply by paambaati** [reply]
@gr2m The above solution worked only once. The workflow is still broken. Is there a possible escape hatch (say force release version) that will help me circumvent this?

**Reply by travi** [reply]
please share more information about the situation. the only reason that this should reoccur is if the history of the release branch was rewritten in some way. the previous release must be in the history of the branch being released from. that means the tag exists and the git notes are intact. do you know of actions that would have resulted in those not existing where they should?

**Reply by travi** [reply]
* https://semantic-release.gitbook.io/semantic-release/usage/configuration#existing-version-tags
* https://semantic-release.gitbook.io/semantic-release/support/troubleshooting#release-not-found-release-branch-after-git-push-force
